tools/wrapgen/main.go

```diff
@@ -0,0 +1,319 @@
+	Methods         []*InterfaceMethod
+	interfacesToAdd []string
+}
+
+type Data struct {
+	// Package is the package name for which to generate code.
+	Package string
+	// Imports contains also _unused_ imports.
+	Imports []*Import
+	// Interface is the interface to generate, including all methods
+	// nested from other interfaces.
+	Interface *Interface
+}
+
+func ExtractParamsList(fl *ast.FieldList) []Declaration {
+	params := make([]Declaration, 0, len(fl.List))
+	for _, field := range fl.List {
+		typeString := types.ExprString(field.Type)
+		for _, n := range field.Names {
+			params = append(params, Declaration{Name: n.Name, Type: typeString})
+		}
+	}
+	return params
+}
+
+func ExtractTypesList(fl *ast.FieldList) []string {
+	typeNames := make([]string, 0, len(fl.List))
+	for _, field := range fl.List {
+		typeString := types.ExprString(field.Type)
+		typeNames = append(typeNames, typeString)
+	}
+	return typeNames
+}
+
+func HasError(t *ast.FieldList) bool {
+	l := t.List
+	if len(l) < 1 {
+		return false
+	}
+	return types.ExprString(l[len(l)-1].Type) == "error"
+}
+
+func GetInterfaceMethod(name string, ft *ast.FuncType) InterfaceMethod {
+	return InterfaceMethod{
+		Name:        name,
+		Params:      ExtractParamsList(ft.Params),
+		ReturnTypes: ExtractTypesList(ft.Results),
+		HasError:    HasError(ft.Results),
+	}
+}
+
+func getImportLocalName(i *ast.ImportSpec) string {
+	if i.Name != nil { // Import renamed in file
+		return strings.Trim(i.Name.Name, "\"")
+	}
+	return ""
+}
+
+var (
+	ErrNotFound       = errors.New("not found")
+	ErrUnexpectedType = errors.New("unexpected type [internal error]")
+)
+
+// flattenNeededInterfaces flattens all methods from interfaces into
+// interfaceName.  Each method is included at most once.
+func flattenNeededInterfaces(interfaces map[string]Interface, interfaceName string) (*Interface, error) {
+	ret := &Interface{Name: interfaceName}
+
+	added := make(map[string]struct{})
+	addMethod := func(m *InterfaceMethod) {
+		if _, ok := added[m.Name]; ok {
+			return
+		}
+		added[m.Name] = struct{}{}
+		ret.Methods = append(ret.Methods, m)
+	}
+
+	// BUG(ariels): If interfaces include each other many times, this
+	//     loop can be very inefficient.
+
+	// todo is a stack of interfaces that still need to be handled.  It
+	// allows for depth-first flattening, which gives a reasonably
+	// readable order of methods.
+	todo := []string{interfaceName}
+
+	for len(todo) > 0 {
+		var nextName string
+		todo, nextName = todo[:len(todo)-1], todo[len(todo)-1]
+		thisInterface, ok := interfaces[nextName]
+		if !ok {
+			return nil, fmt.Errorf("interface %s %w", nextName, ErrNotFound)
+		}
+		for _, method := range thisInterface.Methods {
+			addMethod(method)
+		}
+		// Add any included interfaces in _reverse_ order: they will
+		// be popped, reversing that order again.
+		for i := len(thisInterface.interfacesToAdd) - 1; i >= 0; i-- {
+			todo = append(todo, thisInterface.interfacesToAdd[i])
```

Comment:
And that kids is a perfect example of how every recursion can be implemented with a stack :)

Reply:
:-)